### PR TITLE
[git-webkit] Prune remotes when fetching

### PR DESCRIPTION
**Pull request by JonWBedard**
#### bdb212d5fe625542d0549f14b7c4d578a919d88a
<pre>
[git-webkit] Prune remotes when fetching
<a href="https://bugs.webkit.org/show_bug.cgi?id=247620">https://bugs.webkit.org/show_bug.cgi?id=247620</a>
rdar://102090699

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git): Add an auto-prune option.
(Git.checkout): Add prune argument.
(Git.fetch): Add prune argument, use config default if not specified.
(Git.pull): Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/checkout.py:
(Checkout.parser): Add --prune argument.
(Checkout.main): Pass prune argument to &apos;checkout&apos;.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull.py:
(Pull.parser): Add --prune argument.
(Pull.main): Pass prune argument to &apos;checkout&apos;.

Canonical link: <a href="https://commits.webkit.org/257514@main">https://commits.webkit.org/257514@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f24541204f007ae8fa6cf7b501fe51c141c8015b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/99228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/8434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/32355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/108620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/103227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/8986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104986 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/8986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/32355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/91727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/102751 "webkitpy-tests (failure)") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/8986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/32355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/2314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/32355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/98157 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/2208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/7279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/32355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2632 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/3683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull.py [hunk 1]
```diff
@@ -32,6 +33,20 @@ class Pull(Command):
     aliases = ['up', 'update']
     help = 'Update the current checkout, synchronize git-svn if configured'
 
+    @classmethod
+    def parser(cls, parser, loggers=None):
+        parser.add_argument(
+            '-i', '--issue', '-b', '--bug', '-r',
+            dest='issue', type=str,
+            help='Number (or name) of the issue or bug to create branch for',
+        )
```
Comment (emw-apple) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull.py]: Where did this come from? Seems unrelated

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py [hunk 2]
```diff
@@ -975,7 +978,7 @@ def checkout(self, argument):
             if not rc:
                 return self.commit()
             if rc == 128:
-                run([self.executable(), 'fetch', name], cwd=self.root_path)
+                run([self.executable(), 'fetch', name] + ([] if prune else ['--prune']), cwd=self.root_path)
```
Comment (emw-apple) [inline Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py]: Is this inverted? 
```suggestion
                run([self.executable(), 'fetch', name] + (['--prune'] if prune else []), cwd=self.root_path)
```

**Review by emw-apple — COMMENTED** [review]
I'm not very familiar with how git-webkit's config logic works. Is the default value of `webkitscmpy.auto-prune` None, or True? If the latter, we should probably change it to avoid accidental data loss.

**Comment by JonWBedard** [comment]
> I'm not very familiar with how git-webkit's config logic works. Is the default value of `webkitscmpy.auto-prune` None, or True? If the latter, we should probably change it to avoid accidental data loss.

This will change it to `True`.

Originally, I had agreed with you that not pruning should be the default behavior to avoid data-loss. But in practice, that makes branch management very difficult for admins because contributors basically never add `--prune` on commands interacting with `origin`. Given how we manage PRs, it's also hard for me to come up with a situation where dataloss is an actual problem, folks will be pushing all changes to their personal `fork`, not `origin`. If an `origin` branch is deleted, it was either because admins are pruning a branch that shouldn't have been public or someone mistakenly pushed a PR branch. In these cases, an automatic prune is very desirable.

**Comment by emw-apple** [comment]
> But in practice, that makes branch management very difficult for admins because contributors basically never add `--prune` on commands interacting with `origin`.

That is one of the classic challenges with git :)

Could you make this only ever prune when fetching origin? That seems reasonable. I'm just concerned about the case where I might delete a branch from my fork (maybe I pushed it accidentally?), and then a follow-up `git webkit` command deletes my unmerged local copy.

**Comment by JonWBedard** [comment]
> > But in practice, that makes branch management very difficult for admins because contributors basically never add `--prune` on commands interacting with `origin`.
> 
> That is one of the classic challenges with git :)
> 
> Could you make this only ever prune when fetching origin? That seems reasonable. I'm just concerned about the case where I might delete a branch from my fork (maybe I pushed it accidentally?), and then a follow-up `git webkit` command deletes my unmerged local copy.

I think it would need to be source remotes in general (so, for example, `security` should also be automatically pruned), but I think it's quite reasonable to not automatically prune forks.

**Review by emw-apple — CHANGES_REQUESTED** [review]
[marking `r-` for testing]

**Comment by ddkilzer** [comment]
In this EWS run:  <https://ews-build.webkit.org/#/builders/71/builds/35318>

We have this starting on line 578 of the `checkout-pull-request` step:

```
Auto packing the repository in background for optimum performance.
See "git help gc" for manual housekeeping.
warning: The last gc run reported the following. Please correct the root cause
and remove .git/gc.log.
Automatic cleanup will not be performed until the file is removed.

warning: There are too many unreachable loose objects; run 'git prune' to remove them.
```

Is this the same issue?  Should we also run `rm .git/gc.log` when we prune (or maybe during `clean-up-git-repo`) so git can clean itself up?

**Comment by JonWBedard** [comment]
> In this EWS run: https://ews-build.webkit.org/#/builders/71/builds/35318
> 
> We have this starting on line 578 of the `checkout-pull-request` step:
> 
> ```
> Auto packing the repository in background for optimum performance.
> See "git help gc" for manual housekeeping.
> warning: The last gc run reported the following. Please correct the root cause
> and remove .git/gc.log.
> Automatic cleanup will not be performed until the file is removed.
> 
> warning: There are too many unreachable loose objects; run 'git prune' to remove them.
> ```
> 
> Is this the same issue? Should we also run `rm .git/gc.log` when we prune (or maybe during `clean-up-git-repo`) so git can clean itself up?

This is not the same issue, this change is more concerned with branches that may be deleted from a remote, rather than the orphaned commit problem we see in PR infrastructure.

**Review by emw-apple — APPROVED** [review]
r+, though still confused about the `issue` argument that was added.

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull.py (hunk 1 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull.py]: Thought I deleted that already, removing it now...
